src/serial.rs

```diff
@@ -0,0 +1,503 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides emulation for Linux serial console.
+//!
+//! This is done by emulating an UART 8250 serial port.
+
+use std::collections::VecDeque;
+use std::io::{Result, Write};
+
+use vmm_sys_util::eventfd::EventFd;
+
+// Register offsets.
+const DATA: u8 = 0; // Receiver and Transmitter registers offset, depending on
+                    // the I/O access type: write -> THR, read -> RBR.
+const IER: u8 = 1;
+const IIR: u8 = 2;
+const LCR: u8 = 3;
+const MCR: u8 = 4;
+const LSR: u8 = 5;
+const MSR: u8 = 6;
+const SCR: u8 = 7;
+const DLAB_LOW: u8 = 0;
+const DLAB_HIGH: u8 = 1;
+
+const LOOP_SIZE: usize = 0x40;
+
+const IER_RDA_BIT: u8 = 0b0000_0001;
+const IER_THR_BIT: u8 = 0b0000_0010;
+const IER_8250_BITS: u8 = 0b0000_1111;
```

Comment:
nit: rename this IER_UART_VALID_BITS as it is the definition of valid bytes of IER.